lib/Microservice.js

```diff
@@ -0,0 +1,141 @@
+const { promisify } = require('util');
+const loopback = require('loopback');
+const boot = require('loopback-boot');
+
+
+/**
+ * Basic Microservice class wrapping a loopback application.
+ *
+ * @type {module.Microservice}
+ */
+
+module.exports = class Microservice {
+
+    constructor(app, bootOptions = {}) {
+        const defaultAppDir = './app';
+        const defaultConfigDir = `${defaultAppDir}/config`;
+
+        this.app = app;
+        this.server = null;
+        this._bootOptions = Object.assign({
+            appRootDir: defaultAppDir,
+            appConfigRootDir: defaultConfigDir,
+            componentRootDir: defaultConfigDir,
+            dsRootDir: defaultConfigDir,
+            env: 'dev',
+            middlewareRootDir: defaultConfigDir,
+            modelsRootDir: defaultConfigDir,
+        }, bootOptions);
+    }
+
+    /**
+     * Stop the webserver of the microservice.
+     *
+     * @returns {Promise.<Microservice>}
+     */
+    async stop() {
+
+        if (!this.server || !this.server.listening) return Promise.reject(new Error('Microservice not running'));
+
+        const server = this.server;
+        // avoid race conditions
+        this.server = null;
+
+        return new Promise((resolve, reject) => {
+            // does not work with promisify
+            server.close((error) => {
+                if (error) {
+                    // if unable to stop the server, reassign it.
+                    this.server = server;
+                    return reject(error);
```

Comment:
i personally dont like the pattern where you return a value inside of an if statement without having an else  afterwards. i've seen too many problems caused by this, mostly after the code code changes later on.

Reply:
That's called an early return and is generally used in refactorings to reduce code complexity. But I agree on that as soon as if statements are nested.

Reply:
Could you give me an example of what problems this causes?

Reply:
someone may remove the return statement when refactoring the code and forget about the missing else statement. i've seen this multiple times.